### PR TITLE
rollback bootloader after setting default snapshot

### LINKED ISSUE #77: Recover old bootloader and generate new predictions when doing a rollback
https://bugzilla.suse.com/show_bug.cgi?id=1222183

### PR DESCRIPTION
**Pull request by TobiPeterG**
basic attempt to fix #77

### CODE REVIEW

File: 10-sdbootutil.snapper [hunk 1]
```diff
@@ -45,10 +45,10 @@ set_default_snapshot()
 
 	if is_transactional; then
 		/usr/bin/sdbootutil add-all-kernels "$num" || :
```
Comment (TobiPeterG) [inline 10-sdbootutil.snapper:47]: Why is this required? Doesn't sdbootutil automatically create the entries when with set-default-snapshot if not configs exist?

Comment (aplanas) [reply 10-sdbootutil.snapper:47]: Can be yes. But the explicit call to add-all-kernels is more informative. In any case the call here seems to be more correct. Making the call in 'set-default-spanshot' if there is not config file seems more like a optional decision.

File: sdbootutil [hunk 2]
```diff
@@ -125,9 +125,15 @@ helpandquit()
 		needs-update
 			   Check whether the bootloader in ESP needs updating
 
+		out-of-sync
+			   Check whether the bootloader in ESP has a different version than on the system
+
 		update
 			    Update the bootloader if it's old
 
+		switch
```
Comment (aplanas) [inline sdbootutil:134]: Would it make sense to move this to 'update', with an extra parameter like '--force' or '--sync')

File: sdbootutil [hunk 3]
```diff
@@ -125,9 +125,15 @@ helpandquit()
 		needs-update
 			   Check whether the bootloader in ESP needs updating
 
+		out-of-sync
```
Comment (aplanas) [inline sdbootutil:128]: I do not like the name, and I think that can be added to 'needs-update', with some '--verbose' option, that shows the current version in the ESP and the version in the appointed snapshot

**Review by aplanas — COMMENTED** [review]
I like the approach, but IMHO one (switch) or both commands can be integrated in the existing ones

File: sdbootutil (hunk 2 above)
Comment (TobiPeterG) [reply sdbootutil]: done

File: sdbootutil (hunk 3 above)
Comment (TobiPeterG) [reply sdbootutil]: done

**Comment by TobiPeterG** [comment]
This makes needs_update way more verbose, personally, I like that.

**Comment by TobiPeterG** [comment]
ping
Can I rebase for merge?

**Comment by aplanas** [comment]
> Can I rebase for merge?

Rebase it, sure. But I will need more time to review :disappointed:

**Comment by TobiPeterG** [comment]
Hey, I finally found the time to rebase, feel free to test and review :)

File: sdbootutil [hunk 4]
```diff
@@ -20,6 +20,7 @@ arg_no_variables=
 arg_no_reuse_initrd=
 arg_no_random_seed=
 arg_portable=
+arg_sync=
```
Comment (aplanas) [inline sdbootutil:23]: Is there a situation when we do not want this to be the default?

File: sdbootutil [hunk 5]
```diff
@@ -1543,6 +1554,17 @@ install_bootloader()
 	update_predictions=1
 }
 
+bootloader_update()
+{
+	local status=0
+	bootloader_needs_update "${1:-$root_snapshot}" || status=$?
+	if [[ $status -eq 0 ]]; then
```
Comment (aplanas) [inline sdbootutil:1561]: `[ "$status" = "0" ]` ? I think this is the idiom used in this codebase

File: sdbootutil [hunk 6]
```diff
@@ -1543,6 +1554,17 @@ install_bootloader()
...
+		install_bootloader "${1:-$root_snapshot}"
+	elif [ -n "$arg_sync" ] && [[ $status -eq 2 ]]; then
```
Comment (aplanas) [inline sdbootutil:1563]: `[ "$status" = "2" ]` ?

File: sdbootutil [hunk 7]
```diff
@@ -1458,10 +1461,18 @@ bootloader_needs_update()
 	nv="$(bootloader_version "$(find_bootloader "$snapshot")")"
 	[ -n "$v" ] || return 1
 	log_info "system version $nv"
-	systemd-analyze compare-versions "$v" lt "$nv" 2>/dev/null || return 1
+	systemd-analyze compare-versions "$v" "$nv" 2>/dev/null
+	local status="$?"
 	bldr_name=$(bootloader_name "$snapshot")
-	log_info "$bldr_name needs to be updated"
-	return 0
+	if [ "$status" == "11" ]; then
```
Comment (aplanas) [inline sdbootutil:1467]: single '=' here and in the nexts?

**Review by aplanas — APPROVED** [review]

**Comment by TobiPeterG** [comment]
Oh thank you, I'm sorry that I didn't have a look at your review yet, I'm sick for quite a while now. :(

I planned on working on sdbootutil, but I was laying in bed instead. I will be though when I'm back on full power :)

**Comment by aplanas** [comment]
Please, take care of yourself!!
